protocol-units/da-sequencer/client/src/lib.rs

```diff
@@ -71,9 +63,108 @@ impl DaSequencerClient {
 		};
 
 		let channel = endpoint.connect().await?;
-		let client = DaSequencerNodeServiceClient::new(channel);
+		Ok(DaSequencerNodeServiceClient::new(channel))
+	}
+
+	/// Reconnects the internal gRPC client.
+	async fn reconnect(&mut self) -> Result<()> {
+		tracing::info!("Reconnecting to {}", self.connection_string);
+		let client = Self::connect(&self.connection_string).await?;
+		self.client = client;
+		Ok(())
+	}
+
+	/// Streams blocks starting from a height, with reconnect and resume.
+	pub async fn stream_read_from_height(
+		&mut self,
+		start_request: StreamReadFromHeightRequest,
+	) -> Result<
+		Pin<Box<dyn Stream<Item = Result<StreamReadFromHeightResponse, Status>> + Send>>,
+		Status,
+	> {
+		let height = {
+			let last = self.last_received_height.lock().unwrap();
+			if let Some(last_h) = *last {
+				tracing::info!("Resuming stream from height: {}", last_h + 1);
+				last_h + 1
+			} else {
+				tracing::info!("Starting stream from requested height: {}", start_request.height);
+				start_request.height
+			}
+		};
+
+		match self
+			.client
+			.stream_read_from_height(StreamReadFromHeightRequest { height })
+			.await
+		{
+			Ok(response) => Ok(Self::wrap_stream_with_height_tracking(
+				response.into_inner(),
+				Arc::clone(&self.last_received_height),
+			)),
+			Err(e) => {
+				tracing::warn!("stream_read_from_height failed, trying reconnect: {e}");
+				self.reconnect()
+					.await
+					.map_err(|e| Status::unavailable(format!("Reconnect failed: {e}")))?;
+
+				let response = self
+					.client
+					.stream_read_from_height(StreamReadFromHeightRequest { height })
+					.await?;
+
+				Ok(Self::wrap_stream_with_height_tracking(
+					response.into_inner(),
+					Arc::clone(&self.last_received_height),
+				))
+			}
+		}
+	}
+
+	/// Wraps a stream to track and store the last received height.
+	fn wrap_stream_with_height_tracking(
+		stream: Streaming<StreamReadFromHeightResponse>,
+		last_received_height: Arc<Mutex<Option<u64>>>,
+	) -> Pin<Box<dyn Stream<Item = Result<StreamReadFromHeightResponse, Status>> + Send>> {
+		let wrapped = unfold((stream, last_received_height), |(mut s, tracker)| async move {
+			match s.message().await {
+				Ok(Some(msg)) => {
+					if let Some(ref blob) = msg.response {
+						if let Some(height) = blob.blob_type.as_ref().and_then(|b| match b {
+							movement_da_sequencer_proto::blob_response::BlobType::Blockv1(
+								inner,
+							) => Some(inner.height),
+							_ => None,
+						}) {
+							*tracker.lock().unwrap() = Some(height);
+						}
+					}
+					Some((Ok(msg), (s, tracker)))
+				}
+				Ok(None) => None,
+				Err(e) => Some((Err(e), (s, tracker))),
```

Comment:
If the stream break or return an error, we should try to reconnect. But it starts to be very complicated.